### ISSUE TITLE
Types are not narrowed by multi-pattern `case`

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/pyright#6492]
**Describe the bug**

A `case` with multiple patterns does not discriminate a union type appropriately.

**Code or Screenshots**

Code sample in [pyright playground](https://pyright-play.net/?enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2ANFACqIkAmAIkgMYwBQX7l5AZwFQAggCEAwgAoGCZm04BKAFxco6qAGtUTZYWJkqAbQBE4iSYC6ajeQBG7APrAkJSrqgCYIG%2BvaQIMBRnV3c9Lx9efiEoFgBRADEZRlYOGBVfLR09IlIKSlN4hKtMphJgELcPCMz-CEDglyrw7x4%2BQWEAcQAJAElkuVSlVQ0slA9cwwKTHt6S0bQACyRKsM9WrjLgKDQSGEc6hsdyKVQEAFd9pjS9cygAH1jEh6hZxSgAWgA%2BdZARjQg5Bg7EWmBQFyuaVM2nGVn%2Boz8ghIUDMkhM8IRoya7igAF4wRDHNdOKZDkFVkx5piESA9ucQCgoNimJlRuwkSiiujWdTmXiCZciVCTGTGqFKdZqTS6QymeKeYiBMiZn1uVLRhQkEqoAA5MAwXrYSgkCAkFCkJhxEDgEBSEwAdRIAHJKJQdnsoORwOdxrAwJh8EFkWBtjBFiRtUxyHABABCEyKTKbcru-aixx2U7gwXEmC3SQvIovN6fH4RDGA4Ggs454UwiUYtkc1EWF4mLmN3ni-k1yEkkUBcnMqnqqC0mD0xnMhVQdnalVzTuYzXavUGo0ms0Wq02u2Ol1u3b4L1gH1MP0BqBBq%2Bh8OR6NxhNAA)

I would expect that in `get_common_b`, `input_dict` is narrowed to `ABC | DEF`, and fetching `"common_field"` would cause no errors.

```python
from typing import Literal, TypedDict

class ABC(TypedDict):
    kind: Literal["ABC"]
    abc_field: str
    common_field: str

class DEF(TypedDict):
    kind: Literal["DEF"]
    def_field: str
    common_field: str

class GHI(TypedDict):
    kind: Literal["GHI"]
    ghi_field: str

def get_common_a(input_dict: ABC | DEF | GHI) -> str:
    match input_dict["kind"]:
        case "ABC":
            field = input_dict["common_field"]
            return field
        case "DEF":
            field = input_dict["common_field"]
            return field
        case "GHI":
            raise NotImplementedError("We'll get around to it one of these days!")
    
def get_common_b(input_dict: ABC | DEF | GHI) -> str:
    match input_dict["kind"]:
        case "ABC" | "DEF":
            field = input_dict["common_field"]
            #       ~~~~~~~~~~~~~~~~~~~~~~~~~~
            # error!
            # Could not access item in TypedDict
            #   "common_field" is not a defined key in "GHI"
            return field
        case "GHI":
            raise NotImplementedError("We'll get around to it one of these days!")
```

**VS Code extension or command-line**

I'm running on Pylance v2023.11.10

**Comment by erictraut** [comment]
Pyright contains special-case logic for this particular case, and that logic doesn't handle unions at the moment. For comparison, mypy doesn't handle this even in the non-union case.

I'll look at whether it's relatively straightforward to add support for unions.

**Comment by erictraut** [comment]
I've added support for union types in this code path. This will be included in the next pyright release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
